pennylane/math/quantum.py

```diff
@@ -859,74 +772,91 @@ def fidelity(state0, state1, check_state=False, c_dtype="complex128"):
-            f"{indices10},{indices0},{indices11}->{target}",
-            np.conj(state1),
-            state0,
-            state1,
-            optimize="greedy",
-        )
+def fidelity_statevector(state0, state1, check_state=False, c_dtype="complex128"):
+    r"""Compute the fidelity for two states (given as state vectors) acting on quantum
+    systems with the same size.
 
-        overlap = np.real(overlap)
-        return overlap
-
-    # First state pure, second state mixed
-    if is_state_vector0 and not is_state_vector1:
-        indices00, indices01 = ("ab", "ac") if batched0 else ("b", "c")
-        indices1 = "abc" if batched1 else "bc"
-        target = "a" if batched0 or batched1 else ""
-        overlap = qml.math.einsum(
-            f"{indices00},{indices1},{indices01}->{target}",
-            np.conj(state0),
-            state1,
-            state0,
-            optimize="greedy",
-        )
+    The fidelity for two pure states given by state vectors :math:`\ket{\psi}` and :math:`\ket{\phi}`
+    is defined as
 
-        overlap = np.real(overlap)
-        return overlap
+    .. math::
+        F( \ket{\psi} , \ket{\phi}) = \left|\braket{\psi, \phi}\right|^2
 
-    # Two mixed states
-    fid = _compute_fidelity(state0, state1)
-    return fid
+    This is faster than calling :func:`pennylane.math.fidelity` on the density matrix
+    representation of pure states.
+
+    .. note::
+        It supports all interfaces (Numpy, Autograd, Torch, Tensorflow and Jax). The second state is coerced
+        to the type and dtype of the first state. The fidelity is returned in the type of the interface of the
+        first state.
+
+    Args:
+        state0 (tensor_like): ``(2**N)`` or ``(batch_dim, 2**N)`` state vector.
+        state1 (tensor_like): ``(2**N)`` or ``(batch_dim, 2**N)`` state vector.
+        check_state (bool): If True, the function will check the validity of both states; that is,
+            the shape and the norm
+        c_dtype (str): Complex floating point precision type.
+
+    Returns:
+        float: Fidelity between the two quantum states.
+
+    **Example**
+
+    Two state vectors can be used as arguments and the fidelity (overlap) is returned, e.g.:
+
+    >>> state0 = [0.98753537-0.14925137j, 0.00746879-0.04941796j]
+    >>> state1 = [0.99500417+0.j, 0.09983342+0.j]
+    >>> qml.math.fidelity(state0, state1)
+    0.9905158135644924
+
+    .. seealso:: :func:`pennylane.math.fidelity` and :func:`pennylane.qinfo.transforms.fidelity`
+
+    """
+    # Cast as a c_dtype array
+    state0 = cast(state0, dtype=c_dtype)
+
+    # Cannot be cast_like if jit
+    if not is_abstract(state0):
+        state1 = cast_like(state1, state0)
+
+    if check_state:
+        _check_state_vector(state0)
+        _check_state_vector(state1)
+
+    # Get dimension of the quantum system and reshape
+    num_indices0 = int(np.log2(qml.math.shape(state0)[-1]))
+    num_indices1 = int(np.log2(qml.math.shape(state1)[-1]))

```

Comment:
it seems you're just taking what already existed, but do we need the `log2` if we're just making sure they're equal?

Reply:
Ah good catch